src/examples/4626-router/Generic4626Router.sol

```diff
@@ -0,0 +1,229 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import {
+    toBeforeSwapDelta,
+    BeforeSwapDelta,
+    BeforeSwapDeltaLibrary
+} from "v4-core/src/types/BeforeSwapDelta.sol";
+import {Currency} from "v4-core/src/types/Currency.sol";
+import {PoolKey} from "v4-core/src/types/PoolKey.sol";
+import {PoolId} from "v4-core/src/types/PoolId.sol";
+import {SafeCast} from "v4-core/src/libraries/SafeCast.sol";
+import {IHooks} from "v4-core/src/interfaces/IHooks.sol";
+import {Hooks} from "v4-core/src/libraries/Hooks.sol";
+import {IPoolManager} from "v4-core/src/interfaces/IPoolManager.sol";
+
+import {BaseHook} from "uniswap-hooks/base/BaseHook.sol";
+import {CurrencySettler} from "uniswap-hooks/utils/CurrencySettler.sol";
+
+import {DeltaResolver} from "v4-periphery/src/base/DeltaResolver.sol";
+
+import {ERC4626, ERC20} from "solmate/src/mixins/ERC4626.sol";
+
+/// @title Generic Router for ERC4626 Token Wrappers
+/// @dev Only supports symmetric ERC4626 Vaults
+contract Generic4626Router is BaseHook {
+    using SafeCast for int256;
+    using SafeCast for uint256;
+
+    error Generic4626Router__NotAllowed();
+    error Generic4626Router__InvalidPoolFee();
+
+    struct PoolDetails {
+        bool isInitialized;
+        bool wrapDirection; // true if zeroToOne, false if oneToZero
+        ERC4626 vault;
+        ERC20 underlying;
+    }
+
+    mapping(PoolId poolId => PoolDetails details) public poolDetails;
+
+    constructor(
+        IPoolManager _manager
+    ) BaseHook(_manager) {}
+
+    function initializePool(
+        ERC4626 vault
+    ) external returns (PoolKey memory poolKey, PoolId poolId) {
+        ERC20 underlying = vault.asset();
+        bool wrapZeroForOne = address(underlying) < address(vault);
+
+        poolKey = PoolKey({
+            currency0: wrapZeroForOne
+                ? Currency.wrap(address(underlying))
+                : Currency.wrap(address(vault)),
+            currency1: wrapZeroForOne
+                ? Currency.wrap(address(vault))
+                : Currency.wrap(address(underlying)),
+            fee: 0,
+            tickSpacing: 1, // Irrelevant
+            hooks: IHooks(address(this))
+        });
+
+        poolId = poolKey.toId();
+
+        poolDetails[poolId] = PoolDetails({
+            isInitialized: true,
+            wrapDirection: wrapZeroForOne,
+            vault: vault,
+            underlying: underlying
+        });
+
+        // TODO: I wonder if pool price is truly irrelevant here
+        poolManager.initialize(poolKey, 2 ** 96);
```

Comment:
feel free to remove the TODO, the pool's pool price is technically irrelevant here

Reply:
and interesting idea is whether or not the swap should set the slot0.sqrtPriceX96 to be equal to the exchange rate -- in case solvers use slot0 as a heuristic for pricing..

Reply:
I think not personally, this largely comes down to how people will see this hook. I also don't think the swaps should change poolPrice because they are not a factor at all.